src/rgw/rgw_rest_s3.cc

```diff
@@ -4080,16 +3396,236 @@ bool rgw::auth::s3::RGWS3V2Extractor::is_time_skew_ok(const utime_t& header_time
   }
 }
 
-std::tuple<Version2ndEngine::Extractor::access_key_id_t,
-           Version2ndEngine::Extractor::signature_t,
-           Version2ndEngine::Extractor::string_to_sign_t>
-rgw::auth::s3::RGWS3V2Extractor::get_auth_data(const req_state* const s) const
+
+static rgw::auth::Completer::cmplptr_t
+null_completer_factory(const boost::optional<std::string>& secret_key)
+{
+  return nullptr;
+}
+
+
+using AWSVerAbstractor = AWSEngine::VersionAbstractor;
+
+std::tuple<AWSVerAbstractor::access_key_id_t,
+           AWSVerAbstractor::signature_t,
+           AWSVerAbstractor::string_to_sign_t,
+           AWSVerAbstractor::signature_factory_t,
+           AWSVerAbstractor::completer_factory_t>
+AWSGeneralAbstractor::get_auth_data(const req_state* const s) const
+{
+  AwsVersion version;
+  AwsRoute route;
+  std::tie(version, route) = discover_aws_flavour(s->info);
+
+  if (version == AwsVersion::V2) {
+    return get_auth_data_v2(s);
+  } else if (version == AwsVersion::V4) {
+    return get_auth_data_v4(s, route == AwsRoute::QUERY_STRING);
+  } else {
+    /* FIXME(rzarzynski): handle anon user. */
+    throw -EINVAL;
+  }
+}
+
+static inline
+std::string v4_signature(const std::string& credential_scope,
+
+                         CephContext* const cct,
+                         const boost::string_ref& secret_key,
+                         const boost::string_ref& string_to_sign)
+{
+  auto signing_key = \
+    rgw::auth::s3::get_v4_signing_key(cct, credential_scope, secret_key);
+
+  auto server_signature = \
+    rgw::auth::s3::get_v4_signature(cct, std::move(signing_key),
+                                    string_to_sign);
+
+  return std::string(server_signature.data(), server_signature.size() - 1);
```

Comment:
i don't see any other callers of `rgw::auth::s3::get_v4_signature()` (correct me if i'm wrong) - can that just return `std::string` instead of `std::array` so you can avoid this conversion?